### PR TITLE
Add fee retain modal warning

### PR DESCRIPTION
**Pull request by lubej**
## Solution

Preserve 5x FEE_AMOUNT, upon wrapping 100% ROSE -> wROSE warning modal. Should not popup, if amount to wrap <0.5ROSE or if user balance has more of 5x FEE_AMOUNT left after the wrap.

- add `jest` dependency for testing, to test the popup condition of wrap modal fee;

## Links

Flow here:
- https://www.loom.com/share/7797c0e124074fb8a7a289c94160170e

Design here:
- https://www.figma.com/file/WDJfsYCDhrfufO0Gi6gSdV/ROSE-(un)wrapper-dApp?type=design&node-id=546-5118&mode=design&t=84SkgwIlLQ5hL2JG-4

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -38,8 +40,10 @@
     "eslint-plugin-react-hooks": "^4.6.0",
     "eslint-plugin-react-refresh": "^0.4.3",
     "ethers": "5.7.2",
+    "jest": "^29.7.0",
     "markdownlint-cli": "^0.37.0",
     "prettier": "3.1.0",
+    "ts-jest": "^29.1.2",
```
Comment (lukaw3d) [inline package.json]: vitest seems more appropriate

File: src/components/WrapFeeWarningModal/index.tsx [hunk 2]
```diff
@@ -0,0 +1,68 @@
+import { FC } from 'react'
```
Comment (lukaw3d) [inline src/components/WrapFeeWarningModal/index.tsx:1]: PR name "Add fee reduction modal warning" is misleading. Fee reserve/retain/padding would be better word choices

File: src/components/WrapFeeWarningModal/index.tsx [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+
+        <div className={classes.wrapFeeWarningModalActions}>
+          <Button
+            className={classes.wrapFeeWarningModalButton}
+            onClick={() => next(estimatedAmountWithDeductedFees)}
+          >
+            <span className={classes.wrapFeeWarningModalButtonText}>Wrap reduced amount</span>
+          </Button>
+
+          <a className={classes.wrapFeeWarningModalFullAmount} href={void 0} onClick={() => next(amount!)}>
```
Comment (lukaw3d) [inline src/components/WrapFeeWarningModal/index.tsx]: `href={void 0}`?

(and are you using wrong element `<a>` just for easier css)

File: src/components/icons/TimesIcon.tsx [hunk 4]
```diff
@@ -0,0 +1,17 @@
+import { FC } from 'react'
+
+export const TimesIcon: FC = () => (
```
Comment (lukaw3d) [inline src/components/icons/TimesIcon.tsx:7]: Is this copied from somewhere? does it have a license?

File: src/utils/number.utils.ts [hunk 5]
```diff
@@ -19,4 +20,30 @@ export abstract class NumberUtils {
...
+    // Check if account balance amount even has enough MULTIPLIER x FEE amount
+    // Edge case 0.051 - would only convert 0.001 if user confirms - same goes for the case bellow
+    if (accountBalanceAmount.sub(multiplierDeductionFee).lte(0)) {
+      return false
+    }
+
+    // Check if amount is greater than MULTIPLIER x FEE amount
+    if (amount.sub(multiplierDeductionFee).lte(0)) {
+      return false
+    }
```
Comment (lukaw3d) [inline src/utils/number.utils.ts]:
```suggestion
    if (accountBalanceAmount.lte(multiplierDeductionFee)) {
      return false
    }

    if (amount.lte(multiplierDeductionFee)) {
      return false
    }
```

File: src/utils/__tests__/number.utils.spec.ts [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+
+    it('should return true when amount plus fee times multiplier is greater than account balance', () => {
+      const amount = utils.parseEther('9.99')
+      const accountBalanceAmount = utils.parseEther('10')
+      const show = shouldShowWrapFeeWarningModal({ fee, amount, accountBalanceAmount })
+
+      expect(show).toEqual(true)
+    })
+
+    it('should return false when amount plus fee times multiplier is less than account balance', () => {
```
Comment (lukaw3d) [inline src/utils/__tests__/number.utils.spec.ts]: I don't think theses are good test case descriptions. It's implementation details written in English. Same applies to some comments inside shouldShowWrapFeeWarningModal.

Instead, These should explain *why* is the code as it is

File: src/components/WrapFeeWarningModal/index.tsx [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+  next: (amount: BigNumber) => void
+}
+
+export const WrapFeeWarningModal: FC<WrapFeeWarningModalProps> = ({ isOpen, closeModal, next }) => {
+  const {
+    state: { amount, estimatedFee },
+  } = useWrapForm()
+  const estimatedFeeDeduction = estimatedFee.mul(WRAP_FEE_DEDUCTION_MULTIPLIER)
+
+  const roseAmount = utils.parseUnits(amount ? utils.formatEther(amount) : '0', 'ether')
```
Comment (lukaw3d) [inline src/components/WrapFeeWarningModal/index.tsx]: This looks odd

**Review by lukaw3d — APPROVED** [review]

File: package.json (hunk 1 above)
Comment (lubej) [reply package.json]: Nice, does look clearer

File: src/components/WrapFeeWarningModal/index.tsx (hunk 3 above)
Comment (lubej) [reply src/components/WrapFeeWarningModal/index.tsx]: href="" Could potentially refresh window, and href="#" adds # to the url, so the safest way to do it in React is to just void it.

Not sure which element you are referring to I should be using? button or span?

File: src/components/icons/TimesIcon.tsx (hunk 4 above)
Comment (lubej) [reply src/components/icons/TimesIcon.tsx:7]: I copied the icon from design(link in PR description), contacted Don, and he is saying that it's from Material icons, which should have Apache 2.0 license. Didn't want to import the whole library just for a few icons.

File: src/components/WrapFeeWarningModal/index.tsx (hunk 7 above)
Comment (lubej) [reply src/components/WrapFeeWarningModal/index.tsx]: Agree, refactored.

File: src/utils/__tests__/number.utils.spec.ts (hunk 6 above)
Comment (lubej) [reply src/utils/__tests__/number.utils.spec.ts]: Kinda hard for me to describe math in plain words, did do my best. If you still don't agree with wording of the comments and test descriptions, do suggest some examples.

File: src/components/WrapFeeWarningModal/index.tsx (hunk 3 above)
Comment (lukaw3d) [reply src/components/WrapFeeWarningModal/index.tsx]: - `href={void 0}` is equal to `href={undefined}` is equal to `<a className onClick>`
- button. Don't use anchors, spans, or divs with onClick

File: src/components/icons/TimesIcon.tsx (hunk 4 above)
Comment (lukaw3d) [reply src/components/icons/TimesIcon.tsx:7]: Link to source in a comment

File: src/utils/number.utils.ts [hunk 8]
```diff
@@ -19,4 +20,38 @@ export abstract class NumberUtils {
...
+    // Account balance has NOT enough amount left for future transactions
+    return amount.add(multiplierDeductionFee).gt(accountBalanceAmount)
+  }
+
+  static toBigNumber(amount: BigNumber | null): BigNumber {
+    if (!amount) {
+      return BigNumber.from(0)
+    }
+
+    return amount
```
Comment (lukaw3d) [inline src/utils/number.utils.ts]: I would not expect a function named "toBigNumber" to do this

File: src/components/WrapFeeWarningModal/index.tsx [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+}
+
+export const WrapFeeWarningModal: FC<WrapFeeWarningModalProps> = ({ isOpen, closeModal, next }) => {
+  const {
+    state: { amount, estimatedFee },
+  } = useWrapForm()
+  const estimatedFeeDeduction = estimatedFee.mul(WRAP_FEE_DEDUCTION_MULTIPLIER)
+
+  const roseAmount = NumberUtils.toBigNumber(amount)
+  const estimatedAmountWithDeductedFees = roseAmount!.sub(estimatedFeeDeduction)
```
Comment (lukaw3d) [inline src/components/WrapFeeWarningModal/index.tsx:23]: Should this not be `balance.sub(estimatedFeeDeduction)`

File: src/utils/__tests__/number.utils.spec.ts (hunk 6 above)
Comment (lukaw3d) [reply src/utils/__tests__/number.utils.spec.ts]: I'll try; but I dont understand the `if (amount.sub(multiplierDeductionFee).lte(0)) {` edge case. How do you get to it, without `amount.add(multiplierDeductionFee).gt(accountBalanceAmount)` being false

Comment (lukaw3d) [reply src/utils/__tests__/number.utils.spec.ts]:
```ts
import { expect, test } from 'vitest'
import { utils } from 'ethers'
import { NumberUtils } from '../number.utils.ts'
import { WRAP_FEE_DEDUCTION_MULTIPLIER } from '../../constants/config.ts'

const { shouldShowWrapFeeWarningModal } = NumberUtils

test('should show modal if user can not afford future transactions', () => {
  expect(shouldShowWrapFeeWarningModal({
    accountBalanceAmount: utils.parseEther('10'),
    amount: utils.parseEther('9.99'),
    fee: utils.parseEther('0.01'),
  })).toEqual(true)
})

test('should skip modal if user can already afford future transactions', () => {
  expect(shouldShowWrapFeeWarningModal({
    accountBalanceAmount: utils.parseEther('40'),
    amount: utils.parseEther('20'),
    fee: utils.parseEther('0.01'),
  })).toEqual(false)
})

test('should skip modal if user can not afford to retain fees for future transactions', () => {
  expect(shouldShowWrapFeeWarningModal({
    accountBalanceAmount: utils.parseEther('0.04'),
    amount: utils.parseEther('0.02'),
    fee: utils.parseEther('0.01'),
  })).toEqual(false)
})

test('do not recommend a negative amount / do not affect tiny amount', () => {
  // Note: this one also passes because user can already afford future transactions.
  expect(shouldShowWrapFeeWarningModal({
    accountBalanceAmount: utils.parseEther('10'),
    amount: utils.parseEther('0.005'),
    fee: utils.parseEther('0.01'),
  })).toEqual(false)

  expect(shouldShowWrapFeeWarningModal({
    accountBalanceAmount: utils.parseEther('0.01').mul(WRAP_FEE_DEDUCTION_MULTIPLIER).add(utils.parseEther('0.003')), // 0.053
    amount: utils.parseEther('0.005'),
    fee: utils.parseEther('0.01'),
  })).toEqual(false)
})
```

Comment (lubej) [reply src/utils/__tests__/number.utils.spec.ts]: This case `amount.sub(multiplierDeductionFee).lte(0)`, is for converting small amounts. Like if you would want to wROSE 0.01 ROSE -> wROSE we don't want to show modal - if that makes sense?

File: src/components/WrapFeeWarningModal/index.tsx (hunk 9 above)
Comment (lubej) [reply src/components/WrapFeeWarningModal/index.tsx:23]: No, this should be the amount, you are trying to convert. To see how much ROSE you are going to actually convert after retained fees.

Comment (lukaw3d) [reply src/components/WrapFeeWarningModal/index.tsx:23]: Yeah, so `balance.sub(estimatedFeeDeduction)`

Comment (lubej) [reply src/components/WrapFeeWarningModal/index.tsx:23]: `balance` represents the total amount of ROSE the wallet owns. Where amount is the number in input, you are trying to wrap to wROSE. Try it, and you will see, currently it is working as expected.

Comment (lukaw3d) [reply src/components/WrapFeeWarningModal/index.tsx:23]: if balance = 10
estimatedFeeDeduction = 0.5
amount = 9.8
I would expect it to recommend 9.5 (and it shows a warning if amount is >9.5)

if balance = 0.8
estimatedFeeDeduction = 0.5
amount = 0.4
I would expect it to recommend 0.3 or ignore. Not -0.1 or ignore

Comment (lubej) [reply src/components/WrapFeeWarningModal/index.tsx:23]: Everything working as expected, as you described.

It won't show negative amounts ever. 

I think what you are missing is this code: https://github.com/oasisprotocol/dapp-wrose/blob/12af0be31418fa2f46722303449095d9b0ee16af/src/utils/number.utils.ts#L24-L48 which won't trigger this modal, in the 2nd case. For small amounts <0.5ROSE, the modal will be skipped.
